fix(transformer-optional-catch-binding): the `unused` binding is not in the correct scope

Dunqing opened the pull request:
Blocked by #5008

In the SemanticBuilder, we insert all CatchClause parameters in the BlockStatement scope, a child-scope of CatchClause

https://github.com/oxc-project/oxc/blob/858f510d59cb9ea29cca36c31fdf037b0d627f2c/crates/oxc_semantic/src/builder.rs#L709-L718

So we should do the same thing in this plugin, but because CatchClause has no parameters, SemanticBuilder doesn't create scope for `CatchClause`. This cause we cannot find the `BlockStatement` scope_id.

This PR has changed to the correct logic. Just to wait #5008 solved

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5066?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5066** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5066?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#5065** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5065?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5064** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5064?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Personally I'm not a fan of the way we move the bindings from catch param up into catch block. I understand it's to make identifying illegal declarations easier, but it's confusing/surprising.

We only have to check for illegal declarations in one place (syntax checker) whereas we have to deal with this non-standard scoping everywhere else. So on balance, in my view the downsides probably outweigh the upsides.

If we add scopes for function bodies (#5017), that'd require a mechanism in syntax checker to know when need to check parent scope for a clashing binding anyway, and could use the same mechanism for `catch` too.

Dunqing commented:
> Personally I'm not a fan of the way we move the bindings from catch param up into catch block. I understand it's to make identifying illegal declarations easier, but it's confusing/surprising.
> 
> We only have to check for illegal declarations in one place (syntax checker) whereas we have to deal with this non-standard scoping everywhere else. So on balance, in my view the downsides probably outweigh the upsides.
> 

Yes, I agree with your point. I am not a fan of the way as well! But so far I haven't come up with a better way to address this

> If we add scopes for function bodies (#5017), that'd require a mechanism in syntax checker to know when need to check parent scope for a clashing binding anyway, and could use the same mechanism for `catch` too.

Yes, I am looking forward we having a good mechanism to solve this. But we're stuck with the status quo for now 🥲.

overlookmotel commented:
#5109 unblocks this.

Dunqing commented:
> #5109 unblocks this.

All tests passed, Thanks!

overlookmotel commented on crates/oxc_transformer/src/es2019/optional_catch_binding.rs:
```diff
@@ -58,8 +58,14 @@ impl<'a> Traverse<'a> for OptionalCatchBinding<'a> {
-            ctx.generate_uid("unused", ctx.scoping.current_scope_id(), SymbolFlags::CatchVariable);
+
+        // Get the BlockStatement scope_id below the CatchClause scope
+        let child_scope_id = ctx.scopes().get_child_ids(clause.scope_id.get().unwrap())[0];
```
I think this is simpler, and also avoid a lookup in `ctx.scopes()`:

```suggestion
        let child_scope_id = clause.body.scope_id.get().unwrap();
```

Also, I have been wondering if we can remove `child_ids` from `ScopeTree` - it's rarely used, and you can also always get child scopes by visiting from the node (like `ChildScopeCollector` does), so it's not necessary to have it. I suspect the cost of building and maintaining `child_ids` is not worth it.

overlookmotel replied:
Also, I'd suggest renaming `child_scope_id` var here to `body_scope_id` - that's more descriptive.

Dunqing replied:
Why didn't I realize I could get the scope_id from inside `clause.body`!

overlookmotel reviewed (commented):
Great to see some of the scope tests being taken off the "fail" list. Everything I've been doing over past 2 days has been just adding more and more to that list. Now the tide is turning!

Dunqing committed:
fix(transformer-optional-catch-binding): the `unused` binding is not in the correct scope (#5066)

Blocked by #5008

In the SemanticBuilder, we insert all CatchClause parameters in the BlockStatement scope, a child-scope of CatchClause

https://github.com/oxc-project/oxc/blob/858f510d59cb9ea29cca36c31fdf037b0d627f2c/crates/oxc_semantic/src/builder.rs#L709-L718

So we should do the same thing in this plugin, but because CatchClause has no parameters, SemanticBuilder doesn't create scope for `CatchClause`. This cause we cannot find the `BlockStatement` scope_id.

This PR has changed to the correct logic. Just to wait #5008 solved
